src/python/pants/backend/python/rules/pytest_coverage.py

```diff
@@ -0,0 +1,415 @@
+    # they test, or if they live in a parallel package structure under a separate
+    # source root, such as tests/python/path/to/package testing src/python/path/to/package.
+
+    # Note in particular that this doesn't work for most of Pants's own tests, as those are
+    # under the top level package 'pants_tests', rather than just 'pants' (although we
+    # are moving towards having tests in the same directories as the sources they test).
+
+    # This heuristic is what is used in V1. If we want coverage data on files tested by tests
+    # under `pants_test/` we will need to specify explicitly which files they are testing using
+    # the `coverage` field in the relevant build file.
+    if hasattr(target, "coverage"):
+        return tuple(sorted(set(target.coverage)))
+    return tuple(
+        sorted(
+            {
+                os.path.dirname(source_root_stripped_source_file_path).replace(
+                    os.sep, "."
+                )  # Turn file paths into package names.
+                for source_root_stripped_source_file_path in source_root_stripped_file_paths
+            }
+        )
+    )
+
+
+def ensure_section(config_parser: configparser.ConfigParser, section: str) -> None:
+    """Ensure a section exists in a ConfigParser."""
+    if not config_parser.has_section(section):
+        config_parser.add_section(section)
+
+
+@dataclass(frozen=True)
+class CoveragercRequest:
+    hydrated_targets: HydratedTargets
+    test_time: bool = False
+
+
+@dataclass(frozen=True)
+class Coveragerc:
+    digest: Digest
+
+
+@rule
+async def construct_coverage_config(
+    source_root_config: SourceRootConfig, coverage_config_request: CoveragercRequest
+) -> Coveragerc:
+    sources = await Get[SourceFiles](
+        AllSourceFilesRequest(
+            (ht.adaptor for ht in coverage_config_request.hydrated_targets),
+            strip_source_roots=False,
+        )
+    )
+    init_injected = await Get[InitInjectedSnapshot](InjectInitRequest(sources.snapshot))
+    source_roots = source_root_config.get_source_roots()
+    # Generate a map from source root stripped source to its source root. eg:
+    #  {'pants/testutil/subsystem/util.py': 'src/python'}
+    # This is so coverage reports referencing /chroot/path/pants/testutil/subsystem/util.py can be mapped
+    # back to the actual sources they reference when generating coverage reports.
+    def source_root_stripped_source_and_source_root(file_name: str) -> Tuple[str, str]:
+        source_root = source_roots.find_by_path(file_name)
+        source_root_path = source_root.path if source_root is not None else ""
+        source_root_stripped_path = file_name[len(source_root_path) + 1 :]
+        return (source_root_stripped_path, source_root_path)
+
+    source_to_target_base = dict(
+        source_root_stripped_source_and_source_root(filename)
+        for filename in sorted(init_injected.snapshot.files)
```

Comment:
IIRC, you mentioned wanting to filter out our injected `__init__.py` files from the output? If that's still true, can you please add this:

```
TODO: extend `InitInjectedSnapshot` to have a field `added_files` so that we can filter out all generated init files from the final report.
```

Reply:
I think we will have to do the filtering in our plugin, or possibly in by directly editing the html/xml. 

If we remove the init files from this map, or from the epr environment we'll get errors at report time that the init files don't exist (coverage will be looking in the wrong place for them, or they won't exist)

Reply:
I can leave the comment anyway, but I'm not sure if implementing the use of it will be possible

Reply:
Got it. Feel free to leave off the comment then :)